### ISSUE TITLE
Focus click problems with toolbar

### ISSUE THREAD

**Issue by ellisonbg** [ipython/ipython#8054]
If the selected cell is not in the optimal scroll position when a toolbar button is clicked, the browser will jump to scroll the selected cell into view after the button is clicked. There are some other focus click bugs that have started to show up as well, but I haven't been able to reproduce them. We need to be *very* careful about this stuff...

**Comment by ellisonbg** [comment]
Hmmm, writing my talk I am seeing other focus bugs. I think the issue is at a fairly low level because these problems are showing up in multiple places.

**Comment by HyperionAnalytics** [comment]
I have a problem with browser jump-scrolling when:
1) select part of code;
2) right-click to copy;

It is especially pronounced in "long" cells or cells partially shown in the browser window.

**Comment by Carreau** [comment]
> If the selected cell is not in the optimal scroll position when a toolbar button is clicked, the browser will jump to scroll the selected cell into view after the button is clicked

Can't do something different, or otherwise the button of the toolbar will be selected, and people complained about that, as then you are neither in command, not in edit mode.

**Comment by ellisonbg** [comment]
@Carreau I disagree. This situation is *extremely* bad. We have made a number of changes to how focus is managed and they are cooperating in a particularly nasty way right now to make the UI basically unusable. Just this morning I realized that part of the problem is that simple entering command mode causes a focus jump. Some of the changes we need to revert:

* We should not be focusing cells when they enter command mode.
* We should not be focusing anything when the toolbar is used.
* We should not be disabling keyboard handling when the toolbar is used. When we made that change I was hesitant, but willing to try it. It simply doesn't work.

Summary: we should almost never manually focus any element on the page. The more we do this, the more the different manual focus events start to "argue" leading to crazy focus jumps. As far as I am concerned this is a complete blocker for 3.1.

**Comment by jdfreder** [comment]
I think it may be worth our time to investigate a stronger reliance on the browser's element focus handling.  It seems like these are all problems that the browser should know how to handle itself and that we should be very minimalistic in our custom focus/unfocus logic.

**Comment by jdfreder** [comment]
- Also, I'm not sure how Phosphor adoption will affect this.  It may make things simpler.

**Comment by rsmith31415** [comment]
@ellisonbg 

Could you describe in which situation changing focus after clicking a button is producing unwanted effects? 

If we don't focus cells after clicking on the toolbar, we would need to search the cell with a border to know  which cell was recently moved/created/pasted/etc and then a click is required to enter insert mode. Is this correct?

**Comment by ellisonbg** [comment]
I spent some time looking at this today and think we need to step back a bit:

* Remove the call to register_events on the entire toolbar.
* Remove most of the calls to focus_cell that were added in the actions.js PR. The only calls to focus_cell should be the ones that were in the original keyboard manager. The actions.js PR included a number of new calls to focus cell that don't belong there.
* Audit the code base for other calls to focus_cell and remove ones that are not there for good reason.

Basically I think we need to go back to where things were closer to the 2.0 release. I recall that all of the problems we now have are *new* since then. What @jdfreder and I found when we were writing all of the keyboard handling stuff is that there should be almost no calls to focus_cell - everytime we tried to put one in somewhere it lead to these problems. The calls we had in the 2.0 release were highly tuned and well understood.

**Comment by minrk** [comment]
> Remove the call to register_events on the entire toolbar.

This means that elements in the toolbar can be focused (e.g. cell type select), where keyboard actions make sense, but will actually be triggering shortcuts. We added the call because this behavior was confusing.

@ellisonbg I've opened #8191 with the first two of your bullets. Can you test that one out, and identify specifically what problems remain that we should fix for 3.1?

**Comment by ellisonbg** [comment]
I am fine putting register events on specific limited elements in the
toolbar such as the cell type selector that have their own keyboard
shortcuts. I will give #8191 a shot...

On Sun, Mar 29, 2015 at 4:58 PM, Min RK <notifications@github.com> wrote:

> Remove the call to register_events on the entire toolbar.
>
> This means that elements in the toolbar can be focused (e.g. cell type
> select), where keyboard actions make sense, but will actually be triggering
> shortcuts. We added the call because this behavior was confusing.
>
> @ellisonbg <https://github.com/ellisonbg> I've opened #8191
> <https://github.com/ipython/ipython/pull/8191> with the first two of your
> bullets. Can you test that one out, and identify specifically what problems
> remain that we should fix for 3.1?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/issues/8054#issuecomment-87499183>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com
